src/mpid/ch4/netmod/ofi/ofi_events.c

```diff
@@ -334,94 +334,103 @@ int MPIDI_OFI_get_huge_event(int vni, struct fi_cq_tagged_entry *wc, MPIR_Reques
-        /* recv_elem->localreq may be freed during done_fn.
-         * Need to backup the handle here for later use with MPIDIU_map_erase. */
-        uint64_t key_to_erase = recv_elem->localreq->handle;
-        recv_elem->wc.len = recv_elem->cur_offset;
-        recv_elem->done_fn(vni, &recv_elem->wc, recv_elem->localreq, recv_elem->event_id);
-        ctrl.type = MPIDI_OFI_CTRL_HUGEACK;
-        mpi_errno =
-            MPIDI_OFI_do_control_send(&ctrl, NULL, 0, recv_elem->remote_info.origin_rank,
-                                      recv_elem->comm_ptr, recv_elem->remote_info.ackreq);
-        MPIR_ERR_CHECK(mpi_errno);
+        if (bytesToGet == 0ULL && recv_elem->chunks_outstanding == 0) {
+            MPIDI_OFI_send_control_t ctrl;
+            /* recv_elem->localreq may be freed during done_fn.
+             * Need to backup the handle here for later use with MPIDIU_map_erase. */
+            uint64_t key_to_erase = recv_elem->localreq->handle;
+            recv_elem->wc.len = recv_elem->cur_offset;
+            recv_elem->done_fn(vni, &recv_elem->wc, recv_elem->localreq, recv_elem->event_id);
+            ctrl.type = MPIDI_OFI_CTRL_HUGEACK;
+            mpi_errno =
+                MPIDI_OFI_do_control_send(&ctrl, NULL, 0, recv_elem->remote_info.origin_rank,
+                                          recv_elem->comm_ptr, recv_elem->remote_info.ackreq);
+            MPIR_ERR_CHECK(mpi_errno);
 
-        MPIDIU_map_erase(MPIDI_OFI_global.huge_recv_counters, key_to_erase);
-        MPL_free(recv_elem);
+            MPIDIU_map_erase(MPIDI_OFI_global.huge_recv_counters, key_to_erase);
+            MPL_free(recv_elem);
 
-        goto fn_exit;
-    }
+            goto fn_exit;
+        }
 
-    int nic = 0;
-    int vni_src = recv_elem->remote_info.vni_src;
-    int vni_dst = recv_elem->remote_info.vni_dst;
-    if (MPIDI_OFI_COMM(recv_elem->comm_ptr).enable_striping) {  /* if striping enabled */
-        MPIDI_OFI_cntr_incr(recv_elem->comm_ptr, vni_src, nic);
-        if (recv_elem->cur_offset >= MPIDI_OFI_STRIPE_CHUNK_SIZE && bytesLeft > 0) {
-            for (nic = 0; nic < MPIDI_OFI_global.num_nics; nic++) {
-                int ctx_idx = MPIDI_OFI_get_ctx_index(recv_elem->comm_ptr, vni_dst, nic);
-                remote_key = recv_elem->remote_info.rma_keys[nic];
-
-                bytesLeft = recv_elem->remote_info.msgsize - recv_elem->cur_offset;
-                if (bytesLeft <= 0) {
-                    break;
+        int nic = 0;
+        int vni_src = recv_elem->remote_info.vni_src;
+        int vni_dst = recv_elem->remote_info.vni_dst;
+        if (MPIDI_OFI_COMM(recv_elem->comm_ptr).enable_striping) {      /* if striping enabled */
+            MPIDI_OFI_cntr_incr(recv_elem->comm_ptr, vni_src, nic);
+            if (recv_elem->cur_offset == MPIDI_OFI_STRIPE_CHUNK_SIZE && bytesLeft > 0) {
+                while (bytesLeft > 0) {
+                    for (nic = 0; nic < MPIDI_OFI_global.num_nics; nic++) {
```

Comment:
This issues all `fi_read` concurrently with the same `recv_elem`. I believe that is questionable.

Reply:
The `recv_elem` contains the `cur_offset`, `recv_buf` and `msg_size` of the original large message. The `fi_read()`'s (for the stripes) are happening on different physical NICs at different offsets of the `recv_buf`.

Reply:
Because there is a single `recv_elem`, all the data in `recv_elem` is the same, and one modification will overwrite another, right?

Reply:
@hzhou - Why do you think it would cause issue?

Reply:
Multiple asynchronous operations accessing the same data, potentially concurrent, don't you think that is an issue?

Reply:
They are happening concurrently but on different buffers through different NICs. I do not see any shared variable that is being updated/accessed concurrently.